File: backend/app/services/audit_service.py
from app.extensions import db
from app.models.audit_log import AuditLog
from app.models.user import User

# Clerk/admin package actions
ACTION_PACKAGE_RECEIVED = "package.received"
ACTION_PACKAGE_RECEIVED_UNIDENTIFIED = "package.received_unidentified"
ACTION_PACKAGE_ASSIGNED = "package.assigned"
ACTION_PACKAGE_UNASSIGNED = "package.unassigned"
ACTION_PACKAGE_STATUS_UPDATED = "package.status_updated"
ACTION_PACKAGE_INVOICE_REQUESTED = "package.invoice_requested"
ACTION_PACKAGE_BILLING_UPDATED = "package.billing_updated"
ACTION_PACKAGE_LABEL_UPDATED = "package.label_updated"
ACTION_PACKAGE_PAYMENT_RECORDED = "package.payment_recorded"
ACTION_SHIPMENT_CREATED = "shipment.created"
ACTION_SHIPMENT_PACKAGE_ADDED = "shipment.package_added"
ACTION_SHIPMENT_PACKAGE_REMOVED = "shipment.package_removed"
ACTION_SHIPMENT_DEPARTED = "shipment.departed"


def log_package_action(
    actor: User | None,
    action: str,
    package_id: str,
    summary: str,
    metadata: dict | None = None,
) -> AuditLog:
    if actor and actor.role not in ("clerk", "admin"):
        raise ValueError("Only clerk or admin actions are logged here")

    entry = AuditLog(
        actor_id=actor.id if actor else None,
        actor_name=actor.full_name if actor else "System",
        actor_role=actor.role if actor else "system",
        action=action,
        entity_type="package",
        entity_id=str(package_id),
        summary=summary,
        metadata_json=metadata or {},
    )
    db.session.add(entry)
    db.session.commit()
    return entry


def log_entity_action(
    actor: User | None,
    action: str,
    entity_type: str,
    entity_id: str,
    summary: str,
    metadata: dict | None = None,
) -> AuditLog:
    if actor and actor.role not in ("clerk", "admin"):
        raise ValueError("Only clerk or admin actions are logged here")

    entry = AuditLog(
        actor_id=actor.id if actor else None,
        actor_name=actor.full_name if actor else "System",
        actor_role=actor.role if actor else "system",
        action=action,
        entity_type=entity_type,
        entity_id=str(entity_id),
        summary=summary,
        metadata_json=metadata or {},
    )
    db.session.add(entry)
    return entry

File: server/src/services/auditService.ts
import { db } from '../db/index.js'
import { auditLogs } from '../db/schema/index.js'
import type { UserRow } from '../db/schema/index.js'
import { userFullName } from '../lib/user.js'

export const ACTION_PACKAGE_RECEIVED = 'package.received'
export const ACTION_PACKAGE_RECEIVED_UNIDENTIFIED = 'package.received_unidentified'
export const ACTION_PACKAGE_ASSIGNED = 'package.assigned'
export const ACTION_PACKAGE_UNASSIGNED = 'package.unassigned'
export const ACTION_PACKAGE_STATUS_UPDATED = 'package.status_updated'
export const ACTION_PACKAGE_INVOICE_REQUESTED = 'package.invoice_requested'
export const ACTION_PACKAGE_BILLING_UPDATED = 'package.billing_updated'
export const ACTION_PACKAGE_LABEL_UPDATED = 'package.label_updated'
export const ACTION_PACKAGE_PAYMENT_RECORDED = 'package.payment_recorded'
export const ACTION_SHIPMENT_CREATED = 'shipment.created'
export const ACTION_SHIPMENT_PACKAGE_ADDED = 'shipment.package_added'
export const ACTION_SHIPMENT_PACKAGE_REMOVED = 'shipment.package_removed'
export const ACTION_SHIPMENT_DEPARTED = 'shipment.departed'

export async function logPackageAction(
  actor: UserRow | null,
  action: string,
  packageId: string,
  summary: string,
  metadata?: Record<string, unknown>,
) {
  if (actor && actor.role !== 'clerk' && actor.role !== 'admin') {
    throw new Error('Only clerk or admin actions are logged here')
  }

  const [entry] = await db
    .insert(auditLogs)
    .values({
      actorId: actor?.id ?? null,
      actorName: actor ? userFullName(actor) : 'System',
      actorRole: actor?.role ?? 'system',
      action,
      entityType: 'package',
      entityId: packageId,
      summary,
      metadataJson: metadata ?? {},
    })
    .returning()

  return entry
}

export async function logEntityAction(
  actor: UserRow | null,
  action: string,
  entityType: string,
  entityId: string,
  summary: string,
  metadata?: Record<string, unknown>,
) {
  if (actor && actor.role !== 'clerk' && actor.role !== 'admin') {
    throw new Error('Only clerk or admin actions are logged here')
  }

  const [entry] = await db
    .insert(auditLogs)
    .values({
      actorId: actor?.id ?? null,
      actorName: actor ? userFullName(actor) : 'System',
      actorRole: actor?.role ?? 'system',
      action,
      entityType,
      entityId,
      summary,
      metadataJson: metadata ?? {},
    })
    .returning()

  return entry
}
